Hashes/MD5.js

```diff
@@ -0,0 +1,221 @@
+    chunks.push(str.slice(i, i + size))
+  }
+  return chunks
+}
+
+/**
+ * Rotates the bits to the left
+ *
+ * @param {number} bits - 32 bit number
+ * @param {int} turns - number of rotations to make
+ * @return {string} - number after bits rotation
+ *
+ * @example
+ *      rotateLeft("1011", 3); // "1101"
+ */
+function rotateLeft(bits, turns) {
+  return (bits << turns) | (bits >>> (32 - turns))
+}
+
+/**
+ * Pre-processes message to feed the algorithm loop
+ *
+ * @param {string} message - message to pre-process
+ * @return {string} - processed message
+ */
+function preProcess(message) {
+  // convert message to binary representation padded to
+  // 8 bits, and add 1
+  let m =
+    message
+      .split('')
+      .map((e) => e.charCodeAt(0))
+      .map((e) => e.toString(2))
+      .map((e) => pad(e, 8))
+      .join('') + '1'
+
+  // extend message by adding empty bits (0)
+  while (m.length % 512 !== 448) {
+    m += '0'
+  }
+
+  m = chunkify(m, 32)
+    .map((e) => toLittleEndian(e))
+    .join('')
+
+  // length of message in binary, padded, and extended
+  // to a 64 bit representation
+  let ml = (message.length * CHAR_SIZE).toString(2)
+  ml = pad(ml, 64)
+
+  return m + ml.slice(32) + ml.slice(0, 32)
+}
+
+/**
+ * Hashes message using MD5 Cryptographic Hash Function
+ *
+ * @see
+ * For more info: https://en.wikipedia.org/wiki/MD5
+ * 
+ * @param {string} message - message to hash
+ * @return {string} - message digest (hash value)
+ */
+function MD5(message) {
+  // Initialize variables:
+  const [a0, b0, c0, d0] = [0, 1, 2, 3]
+  const hashes = Uint32Array.from([
+    0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476
+  ])
+
+  // pre-process message and split into 512 bit chunks
+  const bits = preProcess(message)
+  const chunks = chunkify(bits, 512)
+
+  chunks.forEach(function (chunk, _) {
+    // break each chunk into 16 32-bit words
+    const words = Uint32Array.from(
+      chunkify(chunk, 32).map((e) => parseInt(e, 2))
+    )
+
+    // initialize variables for this chunk
+    const [A, B, C, D] = [0, 1, 2, 3]
+    const abcd = Uint32Array.from([
+      hashes[a0],
+      hashes[b0],
+      hashes[c0],
+      hashes[d0]
+    ])
+
+    for (let i = 0; i < 64; i++) {
+      const [F, g] = [0, 1]
+      const fg = Uint32Array.from([0, 0])
+
+      if (i <= 15) {
+        fg[F] = (abcd[B] & abcd[C]) | (~abcd[B] & abcd[D])
+        fg[g] = i
+      } else if (i <= 31) {
+        fg[F] = (abcd[D] & abcd[B]) | (~abcd[D] & abcd[C])
+        fg[g] = (5 * i + 1) % 16
+      } else if (i <= 47) {
+        fg[F] = abcd[B] ^ abcd[C] ^ abcd[D]
+        fg[g] = (3 * i + 5) % 16
+      } else {
+        fg[F] = abcd[C] ^ (abcd[B] | ~abcd[D])
+        fg[g] = (7 * i) % 16
+      }
+
+      fg[F] = fg[F] + abcd[A] + K[i] + words[fg[g]]
+      abcd[A] = abcd[D]
+      abcd[D] = abcd[C]
+      abcd[C] = abcd[B]
+      abcd[B] = abcd[B] + rotateLeft(fg[F], S[i])
+    }
+
+    // add values for this chunk to main hash variables (unsigned)
```

Comment:
1. Use `+=`
2. Shouldn't this be something like wrapping addition? What happens if these grow beyond 32 bits?